# Random sample
Just check random permutations\
(+) can be found global optimum for pure chance, but probability is too low\
(+) super fast\
(-) probably cannot find the best desicion

In [1]:
import pandas as pd
import numpy as np
from tqdm import tqdm
import numpy as np
import random

In [38]:
df = pd.read_csv('/kaggle/input/santa-2024/sample_submission.csv')

In [39]:
df

id                                               text
0   0  advent chimney elf family fireplace gingerbrea...
1   1  advent chimney elf family fireplace gingerbrea...
2   2  yuletide decorations gifts cheer holiday carol...
3   3  yuletide decorations gifts cheer holiday carol...
4   4  hohoho candle poinsettia snowglobe peppermint ...
5   5  advent chimney elf family fireplace gingerbrea...

In [20]:
from metric import PerplexityCalculator
scorer = PerplexityCalculator('/kaggle/input/gemma-2/transformers/gemma-2-9b/2')

Loading checkpoint shards:   0%|          | 0/8 [00:00<?, ?it/s]

Perplexity for raw texts as is

In [5]:
scores = []

for idx, row in tqdm(df.iterrows(), total=len(df)):
    score = scorer.get_perplexity(row.text)

    print(f"Row {idx}: {row.text}")
    print(f"Perplexity Score: {score:.2f}\n")

    scores.append(score)

mean_score = np.mean(scores)
print(f"Mean Perplexity Score: {mean_score:.2f}")

 33%|███▎      | 2/6 [00:00<00:01,  2.46it/s]

Row 0: advent chimney elf family fireplace gingerbread mistletoe ornament reindeer scrooge
Perplexity Score: 3887.90

Row 1: advent chimney elf family fireplace gingerbread mistletoe ornament reindeer scrooge walk give jump drive bake the sleep night laugh and
Perplexity Score: 6068.93



 67%|██████▋   | 4/6 [00:01<00:00,  4.40it/s]

Row 2: yuletide decorations gifts cheer holiday carol magi nutcracker polar grinch sleigh chimney workshop stocking ornament holly jingle beard naughty nice
Perplexity Score: 1118.26

Row 3: yuletide decorations gifts cheer holiday carol magi nutcracker polar grinch sleigh chimney workshop stocking ornament holly jingle beard naughty nice sing cheer and of the is eat visit relax unwrap
Perplexity Score: 1287.11



100%|██████████| 6/6 [00:01<00:00,  4.02it/s]

Row 4: hohoho candle poinsettia snowglobe peppermint eggnog fruitcake chocolate candy puzzle game doll toy workshop wonder believe dream hope peace joy merry season greeting card wrapping paper bow fireplace night cookie milk star wish wreath angel the to of and in that have it not with as you from we kaggle
Perplexity Score: 353.25

Row 5: advent chimney elf family fireplace gingerbread mistletoe ornament reindeer scrooge walk give jump drive bake the sleep night laugh and yuletide decorations gifts cheer holiday carol magi nutcracker polar grinch sleigh chimney workshop stocking ornament holly jingle beard naughty nice sing cheer and of the is eat visit relax unwrap hohoho candle poinsettia snowglobe peppermint eggnog fruitcake chocolate candy puzzle game doll toy workshop wonder believe dream hope peace joy merry season greeting card wrapping paper bow fireplace night cookie milk star wish wreath angel the to of and in that have it not with as you from we kaggle
Perplexity Score: 35

In [6]:
# scorer.clear_gpu_memory()

Mean perplexity for all raw texts id 2178.35   
The worst perplexity has text1 is 6068.93 and text0 is 3887.90     
Let's randomly rearrange this texts just for check, maybe we can improve it a little bit

In [51]:
el = df['text'][0].split()
random.seed(0)
random.shuffle(el)
df.loc[0, 'text'] = ' '.join(el)

In [53]:
el = df['text'][1].split()
random.seed(0)
random.shuffle(el)
df.loc[1, 'text'] = ' '.join(el)

In [54]:
scores = []

for idx, row in tqdm(df.iterrows(), total=len(df)):
    score = scorer.get_perplexity(row.text)

    print(f"Row {idx}: {row.text}")
    print(f"Perplexity Score: {score:.2f}\n")

    scores.append(score)

mean_score = np.mean(scores)
print(f"Mean Perplexity Score: {mean_score:.2f}")

 33%|███▎      | 2/6 [00:00<00:00,  7.28it/s]

Row 0: ornament reindeer chimney gingerbread family fireplace elf advent scrooge mistletoe
Perplexity Score: 2105.57

Row 1: walk laugh sleep bake advent night give elf family scrooge gingerbread ornament fireplace and mistletoe the reindeer chimney drive jump
Perplexity Score: 2833.36



 67%|██████▋   | 4/6 [00:00<00:00,  7.35it/s]

Row 2: yuletide decorations gifts cheer holiday carol magi nutcracker polar grinch sleigh chimney workshop stocking ornament holly jingle beard naughty nice
Perplexity Score: 1118.26

Row 3: yuletide decorations gifts cheer holiday carol magi nutcracker polar grinch sleigh chimney workshop stocking ornament holly jingle beard naughty nice sing cheer and of the is eat visit relax unwrap
Perplexity Score: 1287.11



100%|██████████| 6/6 [00:00<00:00,  7.20it/s]

Row 4: hohoho candle poinsettia snowglobe peppermint eggnog fruitcake chocolate candy puzzle game doll toy workshop wonder believe dream hope peace joy merry season greeting card wrapping paper bow fireplace night cookie milk star wish wreath angel the to of and in that have it not with as you from we kaggle
Perplexity Score: 353.25

Row 5: advent chimney elf family fireplace gingerbread mistletoe ornament reindeer scrooge walk give jump drive bake the sleep night laugh and yuletide decorations gifts cheer holiday carol magi nutcracker polar grinch sleigh chimney workshop stocking ornament holly jingle beard naughty nice sing cheer and of the is eat visit relax unwrap hohoho candle poinsettia snowglobe peppermint eggnog fruitcake chocolate candy puzzle game doll toy workshop wonder believe dream hope peace joy merry season greeting card wrapping paper bow fireplace night cookie milk star wish wreath angel the to of and in that have it not with as you from we kaggle
Perplexity Score: 35

We lower perplexity just with random rearranging words   
We can use it as first level baseline

In [55]:
submission = pd.DataFrame(columns=['id', 'text'])

for idx, row in df.iterrows():

    id = row.id
    optimized_text = row.text

    submission.loc[idx] = {
        'id': id,
        'text': optimized_text
    }

submission.to_csv("submission_random.csv", index=False)

print('submission created')

submission created
